numpy/f2py/src/fortranobject.c

```diff
@@ -388,85 +418,113 @@ fortran_getattr(PyFortranObject *fp, char *name) {
-                if ((PyObject*)arr!=v) {
+        if (fp->defs[i].data != NULL) {
+            /* copy Python object to Fortran array */
+            npy_intp s = PyArray_MultiplyList(fp->defs[i].dims.d,
+                                              PyArray_NDIM(arr));
+            if (s == -1)
+                s = PyArray_MultiplyList(PyArray_DIMS(arr), PyArray_NDIM(arr));
+            if (s < 0 || (memcpy(fp->defs[i].data, PyArray_DATA(arr),
+                                 s * PyArray_ITEMSIZE(arr))) == NULL) {
+                if ((PyObject *)arr != v) {
                     Py_DECREF(arr);
                 }
                 return -1;
             }
-            if ((PyObject*)arr!=v) {
+            if ((PyObject *)arr != v) {
                 Py_DECREF(arr);
             }
-        } else return (fp->defs[i].func==NULL?-1:0);
+        }
+        else {
+            return (fp->defs[i].func == NULL ? -1 : 0);
+        }
         return 0; /* successful */
     }
     if (fp->dict == NULL) {
         fp->dict = PyDict_New();
-        if (fp->dict == NULL)
+        if (fp->dict == NULL) {
             return -1;
+        }
     }
     if (v == NULL) {
         int rv = PyDict_DelItemString(fp->dict, name);
         if (rv < 0)
-            PyErr_SetString(PyExc_AttributeError,"delete non-existing fortran attribute");
+            PyErr_SetString(PyExc_AttributeError,
+                            "delete non-existing fortran attribute");
         return rv;
     }
     else
         return PyDict_SetItemString(fp->dict, name, v);
 }
 
-static PyObject*
-fortran_call(PyFortranObject *fp, PyObject *arg, PyObject *kw) {
+static PyObject *
+fortran_call(PyFortranObject *fp, PyObject *arg, PyObject *kw)
+{
     int i = 0;
     /*  printf("fortran call
         name=%s,func=%p,data=%p,%p\n",fp->defs[i].name,
         fp->defs[i].func,fp->defs[i].data,&fp->defs[i].data); */
-    if (fp->defs[i].rank==-1) {/* is Fortran routine */
-        if (fp->defs[i].func==NULL) {
+    if (fp->defs[i].rank == -1) { /* is Fortran routine */
+        if (fp->defs[i].func == NULL) {
             PyErr_Format(PyExc_RuntimeError, "no function to call");
             return NULL;
         }
-        else if (fp->defs[i].data==NULL)
-            /* dummy routine */
-            return (*((fortranfunc)(fp->defs[i].func)))((PyObject *)fp,arg,kw,NULL);
-        else
-            return (*((fortranfunc)(fp->defs[i].func)))((PyObject *)fp,arg,kw,
-                                                        (void *)fp->defs[i].data);
+        else {
+            if
+        }
+        (fp->defs[i].data == NULL)
+                /* dummy routine */
+                return (*((fortranfunc)(fp->defs[i].func)))((PyObject *)fp,
+                                                            arg, kw, NULL);
+        else return (
+                *((fortranfunc)(fp->defs[i].func)))((PyObject *)fp, arg, kw,
+                                                    (void *)fp->defs[i].data);
```

Comment:
This is really messed up. I'd put some brackets in the original to fix it up.

Reply:
With brackets it forms the (still) ugly:
```c
                return ((*((fortranfunc)(fp->defs[i].func)))((PyObject *)fp,
                                                             arg, kw, NULL));
        else return ((
                *((fortranfunc)(fp->defs[i].func)))((PyObject *)fp, arg, kw,
                                                    (void *)fp->defs[i].data));

```